### PR TITLE
DEV-36065: New PublicApi Java sample

### PR DESCRIPTION
**Pull request by pszymus**

### CODE REVIEW

File: src/java/nps/id/publicapi/java/client/AppListener.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+            // Delivery areas
+            SubscribeDeliveryAreas(edgeUiClient);
+            SubscribeDeliveryAreas(middlewareUiClient);
+
+            // Contracts
+            SubscribeContracts(edgeUiClient, PublishingMode.CONFLATED);
+            SubscribeContracts(middlewareUiClient, PublishingMode.CONFLATED);
+
+            SubscribeContracts(edgeUiClient, PublishingMode.STREAMING);
+            SubscribeContracts(middlewareUiClient, PublishingMode.STREAMING);
```
Comment (paveldayneko) [inline src/java/nps/id/publicapi/java/client/AppListener.java:56]: no need to consume streaming and conflated data at the same time as well

File: src/java/nps/id/publicapi/java/client/AppListener.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            SubscribeCapacities(edgeUiClient, PublishingMode.CONFLATED);
+            SubscribeCapacities(middlewareUiClient, PublishingMode.CONFLATED);
+
+            SubscribeCapacities(edgeUiClient, PublishingMode.STREAMING);
+            SubscribeCapacities(middlewareUiClient, PublishingMode.STREAMING);
+
+            // Order
+            // We wait some time in hope to get some example contracts and configurations that are needed for preparing example order request
+            Thread.sleep(10000);
+            SendOrderRequest(edgeUiClient);
```
Comment (paveldayneko) [inline src/java/nps/id/publicapi/java/client/AppListener.java:72]: PMD can't be used to send orders

File: src/java/nps/id/publicapi/java/client/AppListener.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+            SendOrderRequest(middlewareUiClient);
+
+            System.out.print("Press 'x' key to continue logout, unsubscribe and close. . . ");
```
Comment (paveldayneko) [inline src/java/nps/id/publicapi/java/client/AppListener.java:75]: This example is a bit limited comparing to the previous one.
We should be able to:

- Consume execution reports
- Send invalid order
- Refresh token

Basically we can keep existing application and only extend it with second PMD connection example.
There is no strict need in overriding it

**Review by paveldayneko — CHANGES_REQUESTED** [review]

File: README.md [hunk 4]
```diff
@@ -39,14 +46,41 @@ You need an access token to publish, install, and delete private, internal, and
...
+
 ## Important considerations ##
 
 The current program is using the Spring library and its Websocket Stomp client. The usage of it is at your own discretion. 
 
-In **com/nordpool/intraday/publicapi/example/service/connection/WebSocketConnector.java** please note  the configuration of maxTextMessageSize. Some messages may be quite large. 
+In **WebSocketConnector.java** please note the configuration of maxTextMessageSize and maxByteMessageSize. 
+Some messages may be quite large. 
+
+The example uses ports 8083/443(secured) for establishing the web socket connection with **Middleware** web service and ports 80/443(secured) for establishing web socket connection with **PMD UI** web service.
```
[the remark this answers is not in the document]
Comment (pszymus) [reply README.md]: Unified all PMD/Edge mentions as PMD

File: README.md [hunk 5]
```diff
@@ -1,33 +1,40 @@
...
 ## Building ##
 
-Before you begin, you should have the library  [https://github.com/NordPool/public-intraday-api](https://github.com/NordPool/public-intraday-api) to be available through GitHub packages. 
+Before you begin, you should have the library [https://github.com/NordPool/public-intraday-api](https://github.com/NordPool/public-intraday-api) to be available through GitHub packages. 
 
-Then you may import the code in your favorite IDE and execute **gradle build**.
+Then you may import the code in your favorite IDE and fetch all dependencies with **mvn clean install**.
 
-In order to connect, **you must edit the src\main\resources\example.properties** file:
+All the revelant variables for connecting are located in **application.properties** and Maven's **~\.m2\settings.xml** configuration files.
```
[the remark this answers is not in the document]
Comment (pszymus) [reply README.md]: Added Github relative links

File: src/java/nps/id/publicapi/java/client/connection/options/EdgeWebSocketOptions.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
+package nps.id.publicapi.java.client.connection.options;
+
+import lombok.Getter;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+@Getter
+@Component
+public class EdgeWebSocketOptions implements WebSocketOptions {
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/java/nps/id/publicapi/java/client/connection/options/EdgeWebSocketOptions.java]: Unified

File: README.md [hunk 7]
```diff
@@ -1,33 +1,40 @@
...
+There is also .NET sample application at [https://github.com/NordPool/public-intraday-api-net-example](https://github.com/NordPool/public-intraday-api-net-example). 
 
 ## Disclaimer ##
 
 We offer the data transfer object code and client code examples to aid the development against Nord Pool's API at no warranty whatsoever. Clients are solely responsible for separately testing and ensuring that interaction with Nord Pool works according to their own standards.
 
 ## Building ##
 
-Before you begin, you should have the library  [https://github.com/NordPool/public-intraday-api](https://github.com/NordPool/public-intraday-api) to be available through GitHub packages. 
+Before you begin, you should have the library [https://github.com/NordPool/public-intraday-api](https://github.com/NordPool/public-intraday-api) to be available through GitHub packages.
```
[the remark this answers is not in the document]
Comment (pszymus) [reply README.md:15]: Updated Readme

File: src/java/nps/id/publicapi/java/client/AppListener.java [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+    @Override
+    public void onApplicationEvent(ContextRefreshedEvent event) {
+        try {
+            var edgeUiClient = CreateClient(WebSocketClientTarget.EDGE);
+            var middlewareUiClient = CreateClient(WebSocketClientTarget.MIDDLEWARE);
+
+            // Delivery areas
+            SubscribeDeliveryAreas(edgeUiClient);
+            SubscribeDeliveryAreas(middlewareUiClient);
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/java/nps/id/publicapi/java/client/AppListener.java]: Removed delivery areas consumption from middleware

File: src/java/nps/id/publicapi/java/client/AppListener.java (hunk 1 above)
Comment (pszymus) [reply src/java/nps/id/publicapi/java/client/AppListener.java]: Left only one streaming method subcription

File: src/java/nps/id/publicapi/java/client/AppListener.java (hunk 2 above)
Comment (pszymus) [reply src/java/nps/id/publicapi/java/client/AppListener.java]: Removed

File: src/java/nps/id/publicapi/java/client/AppListener.java (hunk 3 above)
Comment (pszymus) [reply src/java/nps/id/publicapi/java/client/AppListener.java]: Added Consume execution reports, sending order request, sending invalid order request, sending order modification request, sending invalid order modification request

**Comment by pszymus** [comment]
Removed unused files/methods and converted from maven to gradle

**Review by PablitoCBR — DISMISSED** [review]
LGTM 👍

File: README.md [hunk 9]
```diff
@@ -1,33 +1,42 @@
...
 ```
+
 These credentials shall be obtained from [idapi@nordpoolgroup.com](mailto:idapi@nordpoolgroup.com) prior to running the example.
 
-Additionally, make sure that all the properties in the file point to correct addresses.
-Finally, either run the program using **gradle bootRun** or run the class PublicApiApplication.java.
+Additionally, make sure that all the other variables in the [application.properties](/src/main/resources/application.properties) file point to correct addresses.
+Finally, run the solution with **gradle run** command or by run a startup class [ApplicationMain.java](/src/main/java/nps/id/publicapi/java/client/ApplicationMain.java) with your Java IDE.
+
+The program will create two parallel connections that targets both: **Market data** web service and old **Trading** web service.
```
Comment (paveldayneko) [inline README.md:34]: > and old **Trading** web service


'old' will sound confusing for a new clients let's not use 'old' term.
 that targets both: **Market data** web service and **Trading** web service

File: README.md [hunk 10]
```diff
@@ -39,14 +48,38 @@ You need an access token to publish, install, and delete private, internal, and
...
 
-In **com/nordpool/intraday/publicapi/example/service/connection/WebSocketConnector.java** please note  the configuration of maxTextMessageSize. Some messages may be quite large. 
+In addition, sending heartbeat task created after connection established and refreshing access token are  defined in the WebSocketConnector. That connector can be found from [WebSocketConnector.java](src/main/java/nps/id/publicapi/java/client/connection/WebSocketConnector.java).
+Heartbeat interval configuration can be found in [application.properties](src/main/resources/application.properties) **market.data.heartbeat.outgoing.interval** property for Market data web service and **trading.heartbeat.outgoing.interval** for Trading web service.
+
+In [WebSocketConnector.java](/src/main/java/nps/id/publicapi/java/client/connection/WebSocketConnector.java) please note the configuration of **maxTextMessageSize** and **maxByteMessageSize**. 
+Some messages may be quite large. 
+
+The example uses ports 8083/443(secured) for establishing the web socket connection with **Trading** web service and ports 80/443(secured) for establishing web socket connection with **Market data** web service.
```
Comment (paveldayneko) [inline README.md:70]: port 8083 is not available, we are exposing 443 only. Please remove any sign of insecure connection from example
trading.web.socket.useSsl must always be true

**Review by paveldayneko — CHANGES_REQUESTED** [review]

File: README.md (hunk 9 above)
Comment (pszymus) [reply README.md:34]: Yes i forgot to remove this while was updating names of web services

File: README.md (hunk 10 above)
Comment (pszymus) [reply README.md:70]: Sure, I assume should do the same with .net example? 
There is also possibility to configure useSsl parameter.

Comment (paveldayneko) [reply README.md:70]: not really, all our connections are secure, this flag does not make much sense, so it should be removed form both examples

**Review by ElenaVolkova — CHANGES_REQUESTED** [review]

File: README.md [hunk 11]
```diff
@@ -1,33 +1,42 @@
...
-Before you begin, you should have the library  [https://github.com/NordPool/public-intraday-api](https://github.com/NordPool/public-intraday-api) to be available through GitHub packages. 
+Before you begin, you should have the library [https://github.com/NordPool/public-intraday-api](https://github.com/NordPool/public-intraday-api) to be available through GitHub packages. 
+For more detailed instruction see **Authenticating to GitHub Packages** section of this document.
+
 
-Then you may import the code in your favorite IDE and execute **gradle build**.
+Then you may import the code in your favorite IDE and build project with **gradle build** command.
 
-In order to connect, **you must edit the src\main\resources\example.properties** file:
+All the relevant variables for connecting are located in [application.properties](/src/main/resources/application.properties) and Gradle's [build.gradle](build.gradle) configuration files.
```
[the remark this answers is not in the document]
Comment (pszymus) [reply README.md:21]: For me link works well, it won't work from code review mode. You need to switch to the branch on repository, then go to the readme.md file and then links seem to working well.

File: src/main/java/nps/id/publicapi/java/client/ApplicationMain.java [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+import org.springframework.context.annotation.PropertySource;
+import org.springframework.scheduling.TaskScheduler;
+import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;
+
+@SpringBootApplication
+@PropertySource({"classpath:application.properties"})
+public class ApplicationMain {
+
+    public static void main(String[] args) {
+        var context = SpringApplication.run(ApplicationMain.class, args);
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/ApplicationMain.java:15]: context variable removed

File: src/main/java/nps/id/publicapi/java/client/connection/messages/Headers.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
+package nps.id.publicapi.java.client.connection.messages;
+
+public final class Headers {
+    private Headers() {}
+
+    public static final String Heartbeat = "heart-beat";
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/connection/messages/Headers.java:6]: Constants naming fixed

File: src/main/java/nps/id/publicapi/java/client/connection/extensions/StompHeadersExtensions.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+        return value.contains("/streaming")
+                ? PublishingMode.STREAMING
+                : PublishingMode.CONFLATED;
+    }
+
+    public static String getSequenceNumber(StompHeaders stompHeaders) {
+        return stompHeaders.getFirst(Headers.Server.SequenceNumber);
+    }
+
+    public static LocalDateTime getSentAt(StompHeaders stompHeaders) {
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/connection/extensions/StompHeadersExtensions.java:28]: Sure, replaced with ZonedDateTime

File: src/main/java/nps/id/publicapi/java/client/connection/subscriptions/requests/SubscriptionRequestBuilder.java [hunk 15]
```diff
@@ -0,0 +1,75 @@
+package nps.id.publicapi.java.client.connection.subscriptions.requests;
+
+import nps.id.publicapi.java.client.connection.enums.PublishingMode;
+
+import java.util.ArrayList;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class SubscriptionRequestBuilder {
+
+    private final static AtomicInteger atomicInteger = new AtomicInteger();
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/connection/subscriptions/requests/SubscriptionRequestBuilder.java:10]: Fixed

File: src/main/java/nps/id/publicapi/java/client/connection/subscriptions/requests/SubscriptionRequest.java [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+
+    public static SubscriptionRequest publicStatistics(String subscriptionId, String user, String version, PublishingMode publishingMode, int deliveryAreaId) {
+        return new SubscriptionRequest(subscriptionId, "publicStatistics", DestinationHelper.composeDestination(user, version, publishingMode, "publicStatistics/" + deliveryAreaId), PublicStatisticRow.class);
+    }
+
+    public static SubscriptionRequest throttlingLimits(String subscriptionId, String user, String version, PublishingMode publishingMode) {
+        return new SubscriptionRequest(subscriptionId, "throttlingLimits", DestinationHelper.composeDestination(user, version, publishingMode, "throttlingLimits"), ThrottlingLimitsMessage.class);
+    }
+
+    public static SubscriptionRequest capacities(String subscriptionId, String user, String version, PublishingMode publishingMode, int deliveryAreaId, ArrayList<Integer> additionalDeliveryAreas) {
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/connection/subscriptions/requests/SubscriptionRequest.java:67]: Fixed

File: src/main/java/nps/id/publicapi/java/client/connection/subscriptions/requests/SubscriptionRequestBuilder.java [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+
+    public SubscriptionRequest createThrottlingLimits(PublishingMode publishingMode) {
+        return  SubscriptionRequest.throttlingLimits(getSubId(), user, version, publishingMode);
+    }
+
+    public SubscriptionRequest createCapacities(PublishingMode publishingMode, int deliveryAreaId) {
+        return createCapacities(publishingMode, deliveryAreaId, new ArrayList<>());
+    }
+
+    public SubscriptionRequest createCapacities(PublishingMode publishingMode, int deliveryAreaId, ArrayList<Integer> additionalDeliveryAreas) {
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/connection/subscriptions/requests/SubscriptionRequestBuilder.java:68]: Fixed

File: src/main/java/nps/id/publicapi/java/client/startup/AppListener.java [hunk 18]
```diff
@@ -0,0 +1,372 @@
...
+
+import java.time.ZonedDateTime;
+import java.util.*;
+import java.util.concurrent.*;
+
+@Service
+public class AppListener implements ApplicationListener<ContextRefreshedEvent> {
+    private static final Logger logger = LogManager.getLogger(AppListener.class);
+
+    private final String _version = "v1";
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/startup/AppListener.java:35]: Fixed

File: src/main/java/nps/id/publicapi/java/client/connection/storage/SimpleCacheStorage.java [hunk 19]
```diff
@@ -0,0 +1,49 @@
+package nps.id.publicapi.java.client.connection.storage;
+
+import java.util.*;
+
+public class SimpleCacheStorage {
+
+    private static volatile SimpleCacheStorage instance;
+    private static final Object mutex = new Object();
+
+    private final Dictionary<String, List<Object>> _data = new Hashtable<>();
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/connection/storage/SimpleCacheStorage.java:10]: Fixed

File: src/main/java/nps/id/publicapi/java/client/connection/messages/StompMessageFactory.java [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+                put(Headers.Client.AuthorizationToken, Collections.singletonList(authToken));
+                put(Headers.Heartbeat, Collections.singletonList("0," + heartbeatOutgoingInterval));
+            }
+        };
+    }
+
+    public static StompHeaders subscribeHeaders(String destination, String id) {
+        return new StompHeaders() {
+            {
+                setDestination(destination);
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/connection/messages/StompMessageFactory.java:27]: Fixed

File: src/main/java/nps/id/publicapi/java/client/security/responses/AccessTokenResponse.java [hunk 21]
```diff
@@ -0,0 +1,12 @@
+package nps.id.publicapi.java.client.security.responses;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class AccessTokenResponse {
+    @JsonProperty("access_token")
+    public String AccessToken;
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/security/responses/AccessTokenResponse.java:7]: Fixed

File: src/main/java/nps/id/publicapi/java/client/connection/WebSocketConnector.java [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+
+
+public class WebSocketConnector {
+    private static final Logger logger = LogManager.getLogger(WebSocketConnector.class);
+
+    private final ObjectMapper objectMapper = new ObjectMapper();
+
+    private static final String SecureWebSocketProtocol = "wss";
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/connection/WebSocketConnector.java:42]: Fixed

File: src/main/java/nps/id/publicapi/java/client/startup/AppListener.java [hunk 23]
```diff
@@ -0,0 +1,372 @@
...
+                marketDataClient.disconnect();
+                tradingClient.disconnect();
+                System.exit(0);
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    private StompClient CreateClient(WebSocketClientTarget clientTarget) {
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/startup/AppListener.java:118]: Fxied

File: src/main/java/nps/id/publicapi/java/client/security/SsoService.java [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+    private final CredentialsOptions credentialsOptions;
+
+    @Getter
+    private String currentAuthToken = null;
+
+    public SsoService(SsoOptions ssoOptions, CredentialsOptions credentialsOptions)
+    {
+        this.ssoOptions = ssoOptions;
+        this.credentialsOptions = credentialsOptions;
+        httpClient = HttpClient.newHttpClient();
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/security/SsoService.java:39]: Configured DI for http client

File: src/main/java/nps/id/publicapi/java/client/connection/StompFrameHandlerImpl.java [hunk 25]
```diff
@@ -0,0 +1,96 @@
+package nps.id.publicapi.java.client.connection;
+
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.MapperFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.type.TypeFactory;
+import lombok.experimental.ExtensionMethod;
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/connection/StompFrameHandlerImpl.java:7]: Removed extension methods and replaced with classic static methods

File: src/main/java/nps/id/publicapi/java/client/connection/extensions/StompHeadersExtensions.java [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+import org.springframework.messaging.simp.stomp.StompHeaders;
+
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+
+public class StompHeadersExtensions {
+    public static boolean isSnapshot(StompHeaders stompHeaders) {
+        var value = stompHeaders.getFirst(Headers.Server.IsSnapshot);
+        return value.equals("true");
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/connection/extensions/StompHeadersExtensions.java:14]: Added null check for searched header

File: src/main/java/nps/id/publicapi/java/client/connection/extensions/StompHeadersExtensions.java [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    public static PublishingMode getPublishingMode(StompHeaders stompHeaders) {
+        var value = stompHeaders.getFirst(Headers.Destination);
+        return value.contains("/streaming")
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/connection/extensions/StompHeadersExtensions.java:19]: Added null check for searched header

File: src/main/java/nps/id/publicapi/java/client/connection/extensions/StompHeadersExtensions.java [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+                : PublishingMode.CONFLATED;
+    }
+
+    public static String getSequenceNumber(StompHeaders stompHeaders) {
+        return stompHeaders.getFirst(Headers.Server.SequenceNumber);
+    }
+
+    public static LocalDateTime getSentAt(StompHeaders stompHeaders) {
+        var value = stompHeaders.getFirst(Headers.Server.SentAt);
+        return LocalDateTime.ofInstant(Instant.ofEpochMilli(Long.parseLong(value)), ZoneId.of("UTC"));
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/connection/extensions/StompHeadersExtensions.java:30]: Added null check for searched header

File: src/main/java/nps/id/publicapi/java/client/connection/StompFrameHandlerImpl.java [hunk 29]
```diff
@@ -0,0 +1,96 @@
...
+
+    private final ObjectMapper objectMapper = new ObjectMapper();
+
+    private final WebSocketClientTarget clientTarget;
+    private final SubscriptionRequest subscriptionRequest;
+
+    public StompFrameHandlerImpl(WebSocketClientTarget clientTarget, SubscriptionRequest subscriptionRequest) {
+        this.clientTarget = clientTarget;
+        this.subscriptionRequest = subscriptionRequest;
+        objectMapper.configure(MapperFeature.ACCEPT_CASE_INSENSITIVE_PROPERTIES, true);
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/connection/StompFrameHandlerImpl.java:35]: ObjectMapper defined in AppConfiguration file and injected in all places

File: src/main/java/nps/id/publicapi/java/client/connection/clients/StompClientFactory.java [hunk 30]
```diff
@@ -0,0 +1,25 @@
...
+
+    private final TaskScheduler heartBeatScheduler;
+    private final SsoService ssoService;
+
+    public StompClientFactory(TaskScheduler heartBeatScheduler, SsoService ssoService) {
+        this.heartBeatScheduler = heartBeatScheduler;
+        this.ssoService = ssoService;
+    }
+
+    public StompClient create(WebSocketClientTarget clientTarget, String clientId, WebSocketOptions webSocketOptions) {
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/connection/clients/StompClientFactory.java:21]: Configured DI for client id

File: src/main/java/nps/id/publicapi/java/client/startup/AppListener.java [hunk 31]
```diff
@@ -0,0 +1,372 @@
...
+
+            // Capacities
+            subscribeCapacities(marketDataClient, PublishingMode.CONFLATED);
+
+            // Order
+            // We wait some time in hope to get some example contracts and configurations that are needed for preparing example order request
+            Thread.sleep(5000);
+            sendOrderRequest(tradingClient);
+            // Wait before order modification request
+            Thread.sleep(5000);
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/startup/AppListener.java:92]: Replaced with Awaitility

File: src/main/java/nps/id/publicapi/java/client/security/SsoService.java [hunk 32]
```diff
@@ -0,0 +1,75 @@
...
+                    .uri(uri)
+                    .setHeader(Headers.ContentType, "application/x-www-form-urlencoded")
+                    .setHeader(Headers.Authorization, "Basic " + new String(encodedAuth))
+                    .build();
+
+            var response = httpClient.send(httpRequest, HttpResponse.BodyHandlers.ofString());
+            currentAuthToken = objectMapper.readValue(response.body(), AccessTokenResponse.class).AccessToken;
+            return currentAuthToken;
+        }
+        catch (Exception e)
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/security/SsoService.java:66]: Replaced with possible exception types

File: src/main/java/nps/id/publicapi/java/client/security/SsoService.java [hunk 33]
```diff
@@ -0,0 +1,75 @@
...
+            var payload = constructPayloadForTokenRequest(credentialsOptions.getUserName(), credentialsOptions.getPassword());
+            var httpRequest = HttpRequest.newBuilder()
+                    .POST(HttpRequest.BodyPublishers.ofString(payload))
+                    .uri(uri)
+                    .setHeader(Headers.ContentType, "application/x-www-form-urlencoded")
+                    .setHeader(Headers.Authorization, "Basic " + new String(encodedAuth))
+                    .build();
+
+            var response = httpClient.send(httpRequest, HttpResponse.BodyHandlers.ofString());
+            currentAuthToken = objectMapper.readValue(response.body(), AccessTokenResponse.class).AccessToken;
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/security/SsoService.java:63]: It has been tested and it is working well, there are no exceptions during response body deserialization and AccessTokenResponse fields are set and accessible.

File: src/main/java/nps/id/publicapi/java/client/connection/storage/SimpleCacheStorage.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/connection/storage/SimpleCacheStorage.java:10]: Replaced with HashMap

File: src/main/java/nps/id/publicapi/java/client/connection/storage/SimpleCacheStorage.java [hunk 34]
```diff
@@ -0,0 +1,49 @@
...
+
+    private SimpleCacheStorage() {}
+
+    public static SimpleCacheStorage getInstance() {
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/connection/storage/SimpleCacheStorage.java:14]: Removed singleton implementation and configured DI for SimpleCacheStorage in AppConfiguration file

File: src/main/java/nps/id/publicapi/java/client/connection/WebSocketConnector.java [hunk 35]
```diff
@@ -0,0 +1,188 @@
...
+        } catch (Exception e) {
+            logger.error("[SESSION:{}] An error occurred during logout command, details: {}", stompSession.getSessionId(), e.getMessage());
+        }
+    }
+
+    public void disconnect() {
+        stompSession.disconnect();
+    }
+
+    private void periodicallyRefreshToken() throws Exception {
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/connection/WebSocketConnector.java:125]: Moved to StompSessionHandlerAdapterImpl class

File: src/main/java/nps/id/publicapi/java/client/startup/AppListener.java [hunk 36]
```diff
@@ -0,0 +1,372 @@
...
+            sendOrderModificationRequest(tradingClient);
+            // Wait before invalid order request
+            Thread.sleep(5000);
+            sendInvalidOrderRequest(tradingClient);
+            // Wait before invalid order modification request
+            Thread.sleep(5000);
+            sendInvalidOrderModificationRequest(tradingClient);
+
+            Thread.sleep(5000);
+            System.out.println("============================================================ ");
```
[the remark this answers is not in the document]
Comment (pszymus) [reply src/main/java/nps/id/publicapi/java/client/startup/AppListener.java:106]: Replaced with Logger

**Review by ElenaVolkova — APPROVED** [review]

**Review by paveldayneko — APPROVED** [review]

**Pull request merged**